### PR TITLE
backport_branch: keep packages that own .link file sources

### LINKED ISSUE #20726: backport_branch.sh: keep packages that are sources of .link files in the target package
## Problem

When creating a backport branch, the script removes all packages except the target package and any packages it declares under `requires.input` / `requires.content` in its `manifest.yml` (handled by `get_required_package_names` in `backport_branch_lib.sh`).

However, packages can contain [`.link` files](https://www.elastic.co/docs/extend/integrations/developer_workflow_links) — plain-text files with a `.link` extension managed by `elastic-package`. Each `.link` file contains a path (relative to the link file itself) pointing to a source file whose content it mirrors, plus a SHA-256 checksum. That relative path can traverse `../` boundaries and reference a file that lives inside a **different package**.

When a linked file's source package is not the target package and is not listed under `requires.*`, `remove_other_packages` will delete it, leaving the `.link` file pointing at a non-existent file and breaking the backport branch.

This was raised during review of [elastic/package-spec#1209](https://github.com/elastic/package-spec/pull/1209#discussion_r3782819412).

## Expected behaviour

`backport_branch.sh` should detect all `*.link` files inside the target package, resolve their source paths, identify which package each source file belongs to (by finding the nearest `manifest.yml` ancestor), and add those packages to `packages_to_keep` — the same list that already preserves `requires.*` dependencies.

`.github/CODEOWNERS` entries for those extra packages must also be retained, which is already handled generically by `remove_other_packages` for packages in `packages_to_keep`, so no extra logic is needed there once the package list is correct.

## Implementation hints

`backport_branch_lib.sh` already has the pattern to follow in `get_required_package_names`. A new function — e.g. `get_linked_source_package_names` — should:

1. Walk all `*.link` files under the given package path (`find "$package_path" -name "*.link"`).
2. Read the first whitespace-delimited field from each file (the relative path to the source).
3. Resolve the absolute path: `dirname(<link_file>)/<relative_path>`.
4. Walk up from that resolved path to find the nearest directory containing a `manifest.yml` — that directory is the source package root.
5. If the source package root differs from the target package root, emit its path so the caller can add it to `packages_to_keep`.

For reference, `elastic-package` implements the equivalent logic in Go at [`internal/files/linkedfiles.go`](https://github.com/elastic/elastic-package/blob/main/internal/files/linkedfiles.go) — specifically `listLinkedFiles` + `newLinkedFile` (which populates `IncludedPackageName`).

## Files to change

- `.buildkite/scripts/backport_branch_lib.sh` — add `get_linked_source_package_names`
- `.buildkite/scripts/backport_branch.sh` — call it alongside `get_required_package_names` when building `packages_to_keep`

### PR DESCRIPTION
**Pull request by mrodm**
## Proposed commit message

```
backport_branch: keep packages that own .link file sources

When backport_branch.sh creates a backport branch it removes all packages
except the target. If the target (or any of its requires.* dependencies)
contains .link files pointing into another package, those source packages
were also deleted — leaving dangling links and a broken branch.

Any package that is referenced as a .link source — directly or
transitively through chains of .link files and requires.* dependencies —
is now treated as a required dependency and retained alongside the target
package.

WHAT:

Two new functions in backport_branch_lib.sh:

- collect_linked_packages_from_roots: BFS over .link files starting from
  a set of root packages. Resolves each .link's relative path with
  realpath -m, matches against a one-time cache of all packages from
  list_all_directories, and emits any package not already in the input
  set. Self-links (source within the same package) are silently skipped.
  Path prefix checks use ${var#"prefix"/} (literal, not glob) to avoid
  mismatches on CI runner paths containing bracket characters.

- collect_packages_to_keep: the main entry point called by
  backport_branch.sh. Iteratively expands a keep-set until stable,
  alternating requires.* expansion and .link BFS. Handles all cross-type
  chaining: target → requires → link (required package has .link files)
  and target → link → requires (link-discovered package has requires.*
  deps).

WHY:

A package that uses .link files to share common field definitions, pipeline
configurations, or other assets from another package will have those source
packages deleted when a backport branch is created for it. For example, if
package A contains .link files referencing files inside package B, the
current code removes B, leaving A with broken links and an unusable
backport branch. The iterative expansion handles cases where a requires.*
dependency also has .link files, or a link-discovered package has its own
requires.* entries — a single pass over either type would miss the other.
```

## Changes

Two new functions are added to `backport_branch_lib.sh`:
`collect_linked_packages_from_roots`, which performs a BFS over `.link`
files from a set of root packages to discover all packages whose files are
referenced; and `collect_packages_to_keep`, the main entry point that
iteratively expands a keep-set by alternating `requires.*` resolution and
`.link` BFS until stable. `backport_branch.sh` is updated to call
`collect_packages_to_keep` instead of building the keep-set inline.
`test_backport_branch.sh` is extended with 57 tests covering unit,
edge-case, and integration scenarios for all expansion paths (direct links,
transitive chains, required packages with links, and link-discovered
packages with requires).

How `collect_packages_to_keep` discovers all packages to retain:

  1. Seed the keep-set with the target package.
  2. For each package not yet checked, read `requires.input[]` and
     `requires.content[]` from its `manifest.yml` and add those packages.
  3. Run BFS from the entire keep-set: for each `.link` file found,
     resolve the source path and identify which package owns it; add
     that package if not already in the keep-set.
  4. Repeat steps 2–3 until no new packages are added (stable).

  Additionally:
  - Self-links (`.link` pointing to a file within the same package) are
    silently skipped at step 3.
  - Transitive chains are handled automatically: a package discovered
    via `.link` in step 3 will have its own `requires.*` and `.link`
    files checked in the next iteration.
  - Sources outside any known package (e.g. `_dev/shared/` at repo root)
    emit a warning on stderr and are skipped without failing the build.

## Author's Checklist

- [ ] Run `bash .buildkite/scripts/test_backport_branch.sh` — 57 tests pass
- [ ] Verify manually with `collect_packages_to_keep packages/windows` against real packages (cross-package links are confirmed present in the repo)

## How to test this PR locally

**Option 1 — Unit tests (no dependencies):**
```bash
bash .buildkite/scripts/test_backport_branch.sh
# Expected: 57 passed, 0 failed
```

**Option 2 — Run against real packages:**
```bash
source .buildkite/scripts/backport_branch_lib.sh
list_all_directories() { mage -d . listPackages; }
collect_packages_to_keep packages/windows
# Expected: packages/windows (all links are self-links unless cross-package ones exist)
```

**Option 3 — Synthetic cross-package link:**
```bash
echo "../../../../azure/_dev/shared/fields/base-fields.yml fakechecksum" \
  > packages/windows/data_stream/applocker_packaged_app_deployment/fields/test.link

source .buildkite/scripts/backport_branch_lib.sh
list_all_directories() { mage -d . listPackages; }
collect_packages_to_keep packages/windows
# Expected: packages/windows + packages/azure

rm packages/windows/data_stream/applocker_packaged_app_deployment/fields/test.link
```

## CI examples

- [Build #335](https://buildkite.com/elastic/integrations-backport/builds/335) — backport of a package that includes a `requires` field (`elastic_package_registry`); all required packages are correctly retained.
- [Build #334](https://buildkite.com/elastic/integrations-backport/builds/334) — backport of `packages/windows`, which contains `.link` files pointing to files within the same package (self-links); no extra packages are retained, confirming self-links are correctly filtered out.

## Related issues

- Closes https://github.com/elastic/integrations/issues/20726

---

> This PR was generated with the assistance of [Claude](https://claude.ai) (claude-sonnet-4-6).

### CODE REVIEW

File: .buildkite/scripts/backport_branch.sh [hunk 1]
```diff
@@ -68,9 +68,9 @@ FULL_ZIP_PACKAGE_NAME="${PACKAGE_NAME}-${PACKAGE_VERSION}.zip"
 TRIMMED_PACKAGE_VERSION="$(echo "$PACKAGE_VERSION" | cut -d '.' -f -2)"
 SOURCE_BRANCH="main"
 ## In order to test other branches probably it is required to copy the dev files or magefile from the PR branch, and for that it would require these changes
-# git checkout -b test_main
-# SOURCE_BRANCH="$(git rev-parse --abbrev-ref HEAD)"
-# echo "--- SOURCE_BRANCH: ${SOURCE_BRANCH}"
+git checkout -b test_main
+SOURCE_BRANCH="$(git rev-parse --abbrev-ref HEAD)"
+echo "--- SOURCE_BRANCH: ${SOURCE_BRANCH}"
```
Comment (mrodm) [inline .buildkite/scripts/backport_branch.sh]: To be reverted before merging

File: .buildkite/scripts/backport_branch_lib.sh [hunk 2]
```diff
@@ -35,6 +37,143 @@ get_required_package_names() {
...
 
+
+# collect_linked_packages_from_roots accepts any number of root package paths
+# and returns (one per line) all packages reachable via .link files from any
+# of those roots, transitively. The roots themselves are excluded. Output is
+# deduplicated across all roots so each package appears at most once.
+# list_all_directories is called exactly once per invocation of this function,
+# regardless of the number of roots or BFS hops. collect_packages_to_keep calls
+# this function once per outer iteration (see its known-limitation comment).
+collect_linked_packages_from_roots() {
```
Comment (teresaromero) [inline .buildkite/scripts/backport_branch_lib.sh]: i am not sure i understand the naming of "roots" here; this collects linked files from the root of the integrations repository? regardless if these are under the packages directory?

File: .buildkite/scripts/backport_branch.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (mrodm) [reply .buildkite/scripts/backport_branch.sh]: To be removed before merging

File: .buildkite/scripts/backport_branch_lib.sh (hunk 2 above)
Comment (mrodm) [reply .buildkite/scripts/backport_branch_lib.sh]: Those roots are referring to the current package paths discovered in this process.

For every new package discovered (for instance a package from a `requires` field), it needs to be checked if that new package whether or not that package contains link files. If there is any package, then those packges must be kept in the backport branch.

In any case, I've renamed the function (and its usages) to try be more descriptive about what it does here  https://github.com/elastic/integrations/pull/21569/commits/05d0e38c15de45503e2ec7c746274181dadeb6ea

File: .buildkite/scripts/backport_branch.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (mrodm) [reply .buildkite/scripts/backport_branch.sh]: Reverted https://github.com/elastic/integrations/pull/21569/commits/55705695a3e199ca2f4850606237434d9ba0c6a2

Comment (mrodm) [reply .buildkite/scripts/backport_branch.sh]: Reverted here https://github.com/elastic/integrations/pull/21569/commits/55705695a3e199ca2f4850606237434d9ba0c6a2

File: .buildkite/scripts/backport_branch_lib.sh [hunk 3]
```diff
@@ -35,6 +37,146 @@ get_required_package_names() {
   yq -r '.requires.content[].package' "${manifest}" 2>/dev/null || true
 }
 
+
+# collect_link_source_packages accepts any number of package paths and returns
+# (one per line) all packages reachable via .link files from any of those
+# packages, transitively. The input packages themselves are excluded. Output is
+# deduplicated so each package appears at most once.
+# Only packages under the packages/ directory (as returned by list_all_directories)
+# are considered as potential link sources. Sources that resolve outside that set
```
Comment (mrodm) [inline .buildkite/scripts/backport_branch_lib.sh:46]: Created an issue https://github.com/elastic/integrations/issues/21594 in case it is needed at some point to support links files outside of packages, and validate if it requires any further action.

Probably, as `remove_other_packges` function won't delete those files from the backport branch, it is not needed to do anything. But it is something to validate/investigate further as part of that issue.

**Review by teresaromero — APPROVED** [review]
🚀  thanks
